Fix misaligned Dropdown panel

rebeccaalpert opened the pull request:
Fixes [issue #154](https://github.com/openshift/console/issues/154).

Adjusted classes on button and outer Dropdown div. Button classes are now applied to the correct div when we're faking a button. Dropdowns with real buttons should be unaffected.

![screen shot 2018-06-25 at 1 28 22 pm](https://user-images.githubusercontent.com/7014965/41865669-c0c0600e-787b-11e8-9475-9111b88a1349.png)

coreos-ui commented:
Can one of the admins verify this patch?

coreos-ui commented:
Can one of the admins verify this patch?

spadgett commented:
jenkins rebuild

spadgett commented:
/assign @rhamilto

rhamilto commented:
I think the actual bug is that we *should* be using a `<button>` in list-page.jsx, clusterserviceversion.tsx, and cluster-pickewr.tsx and not a faked one (note: masthead.tsx should remain a faked button).  @spadgett?

spadgett commented:
Agreed, we should be using real `<button>` elements when we can for accessibility. We should switch them if it's not too difficult.

rebeccaalpert commented:
I switched the dropdowns on list-page.jsx, clusterserviceversion.tsx, and cluster-picker.tsx so they use buttons. I added an extra prop to Dropdown so I could address them specifically (since they needed classes added to override the default button styles) and adjusted the styling on ClusterPicker to accommodate its new structure.

rebeccaalpert commented:
I made changes in response to your feedback; let me know if you see anything else.

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -289,15 +289,18 @@ export class Dropdown extends DropdownMixin {
-      ? <div onClick={this.toggle} className="dropdown__not-btn" id={this.props.id}>
+    let button;
+    if (noButton) {
+      button = <div onClick={this.toggle} className="dropdown__not-btn" id={this.props.id}>
```
Ideally this would be a button element anyway for accessibility, just without the button styling. Not asking you to fix that in this PR... Just noting the existing dropdown is not keyboard or screen reader accessible.

We should open a JIRA issue.

rebeccaalpert commented:
I addressed the other feedback in my new commit.

rhamilto commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -289,11 +289,11 @@ export class Dropdown extends DropdownMixin {
 
-    const button = noButton
-      ? <div onClick={this.toggle} className="dropdown__not-btn" id={this.props.id}>
+    const button = (noButton) ?
```
Our typical pattern is to put the ? and the : on the same line as the output.  Making this change will reduce the diff.

rhamilto replied:
And omitting the parentheses around `noButton` will further reduce the diff.

rebeccaalpert commented:
All set.

rhamilto commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -292,8 +292,7 @@ export class Dropdown extends DropdownMixin {
         <span className="dropdown__not-btn__title">{title}</span>&nbsp;<Caret />
-      </div>
-      : <button onClick={this.toggle} type="button" className="btn btn-default btn--dropdown dropdown-toggle" id={this.props.id}>
+      </div> : <button onClick={this.toggle} type="button" className={`btn btn-default btn--dropdown dropdown-toggle ${className}`} id={this.props.id}>
```
Ideally the colon and everything after is on its own line so the formatting is consistent across the code and the only resulting diff is on line 296.
```
</div> 
: <button onClick={this.toggle} type="button" className={`btn btn-default btn--dropdown dropdown-toggle ${className}`} id={this.props.id}>
```

rebeccaalpert commented:
All set.

rhamilto commented:
Unfortunately, the positioning of .co-m-pane__filter-bar-group .dropdown-menu is now incorrect. 

![screen shot 2018-07-11 at 9 07 03 am](https://user-images.githubusercontent.com/895728/42573710-d5aabb7e-84ea-11e8-8f6c-8ca22f4be5c0.PNG)

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -293,7 +293,7 @@ export class Dropdown extends DropdownMixin {
         <span className="dropdown__not-btn__title">{title}</span>&nbsp;<Caret />
       </div>
-      : <button onClick={this.toggle} type="button" className="btn btn-default btn--dropdown dropdown-toggle" id={this.props.id}>
+      : <button onClick={this.toggle} type="button" className={`btn btn-default btn--dropdown dropdown-toggle ${className}`} id={this.props.id}>
```
I think it's better to have separate props for button className and dropdown className. It's confusing that it's used in different ways depending on `props.noButton`

rebeccaalpert commented:
I reworked this to use separate props for button className and dropdown className. I checked the dropdowns I can see and they look fine. Let me know if you have feedback.

rhamilto commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -323,9 +313,19 @@ export class Dropdown extends DropdownMixin {
+            ? <div onClick={this.toggle} className="dropdown__not-btn" id={this.props.id}>
+              <span className="dropdown__not-btn__title">{title}</span>&nbsp;<Caret />
+            </div>
+            : <button onClick={this.toggle} type="button" className={classNames('btn', 'btn-default', 'btn--dropdown', 'dropdown-toggle', {[buttonClassName]: buttonClassName})} id={this.props.id}>
```
I think it'd be better to do `classNames('btn', 'btn--dropdown', 'dropdown-toggle', buttonClassName ? buttonClassName : 'btn-default')` so that btn-default and {buttonClassName} aren't both added.

rhamilto commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -323,9 +313,19 @@ export class Dropdown extends DropdownMixin {
 
     //Adding `dropDownClassName` specifically to use patternfly's context selector component, which expects `bootstrap-select` class on the dropdown. We can remove this additional property if that changes in upcoming patternfly versions.
-    return <div className={className} ref={this.dropdownElement} style={this.props.style}>
+    return <div className={classNames({[wrapperClassName]: wrapperClassName})} ref={this.dropdownElement} style={this.props.style}>
```
I believe you can just do `className={wrapperClassName}` here.

rhamilto commented on frontend/public/components/namespace.jsx:
```diff
@@ -249,9 +249,9 @@ class NamespaceDropdown_ extends React.Component {
+        wrapperClassName="co-namespace-selector__dropdown"
         menuClassName="co-namespace-selector__menu"
-        noButton
+        noButton={true}
```
Nit:  no need to declare a value.  If the prop exists, the value is true.

rhamilto commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -387,6 +387,7 @@ export const ActionsMenu = (props) => {
     }
   };
   return <Dropdown className="btn--actions"
+    dropdownClass={true}
```
Orphaned prop?

rhamilto commented on frontend/public/components/resource-dropdown.tsx:
```diff
@@ -109,7 +109,7 @@ const ResourceListDropdown_: React.StatelessComponent<ResourceListDropdownProps>
     items[selected] = <DropdownItem kind={selected} />;
   }
-  return <Dropdown className={classNames('co-type-selector', {[className]: className})} items={items} title={items[selected]} onChange={onChange} selectedKey={selected} />;
+  return <Dropdown wrapperClassName={classNames('co-type-selector', {[className]: className})} items={items} title={items[selected]} onChange={onChange} selectedKey={selected} />;
```
I believe you can just do `classNames('co-type-selector', className)` here.

spadgett commented:
/hold

Jenkins is down so integration tests aren't running

spadgett commented:
jenkins rebuild

spadgett commented:
/hold cancel

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -435,7 +436,7 @@ export class ContainerDropdown extends React.PureComponent {
     const title = _.get(dropdownItems, currentKey) || containerLabel(firstContainer);
     return <Dropdown
-      className="btn-group"
+      wrapperClassName="btn-group"
```
`wrapperClassName` is a bit confusing. It's not clear what the "wrapper" element really is without looking at the dropdown source.

I'd probably just keep `className` for the dropdown itself and `buttonClassName` for the button. (The important thing is to have a separate prop for the child button class.)

rebeccaalpert commented:
I changed wrapperClassName to className.

spadgett commented:
/lgtm

alecmerdler commented:
`GUI-Tests-ALM` are failing due to a changed CSS selector.

rebeccaalpert commented:
I modified the ALM integration tests.

spadgett commented:
/lgtm

rebeccaalpert commented:
/retest

rebeccaalpert commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

spadgett commented:
ALM tests should be working again

rebeccaalpert commented:
jenkins rebuild

rebeccaalpert commented:
/retest

rebeccaalpert commented:
jenkins rebuild
